modules/merge/merge.go

```diff
@@ -0,0 +1,479 @@
+		}
+
+		// Set custom message and author and create merge commit
+		sig := doer.NewGitSig()
+		if err := git.NewCommand("commit", fmt.Sprintf("--author='%s <%s>'", sig.Name, sig.Email), "-m", message).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
+			return fmt.Errorf("git commit [%s]: %v - %s", tmpBasePath, err, errbuf.String())
+		}
+
+	case models.MergeStyleSquash:
+		// Merge with squash
+		if err := git.NewCommand("merge", "-q", "--squash", trackingBranch).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
+			return fmt.Errorf("git merge --squash [%s -> %s]: %s", headRepoPath, tmpBasePath, errbuf.String())
+		}
+		sig := pr.Issue.Poster.NewGitSig()
+		if err := git.NewCommand("commit", fmt.Sprintf("--author='%s <%s>'", sig.Name, sig.Email), "-m", message).RunInDirPipeline(tmpBasePath, nil, &errbuf); err != nil {
+			return fmt.Errorf("git commit [%s]: %v - %s", tmpBasePath, err, errbuf.String())
+		}
+	default:
+		return models.ErrInvalidMergeStyle{ID: pr.BaseRepo.ID, Style: mergeStyle}
+	}
+
+	if setting.LFS.StartServer {
+		// Now we have to implement git lfs pre-push
+		// git rev-list --objects --filter=blob:limit=1k HEAD --not base
+		// pass blob shas in to git cat-file --batch-check (possibly unnecessary)
+		// ensure only blobs and <=1k size then pass in to git cat-file --batch
+		// to read each sha and check each as a pointer
+		// Then if they are lfs -> add them to the baseRepo
+		revListReader, revListWriter := io.Pipe()
+		shasToCheckReader, shasToCheckWriter := io.Pipe()
+		catFileCheckReader, catFileCheckWriter := io.Pipe()
+		shasToBatchReader, shasToBatchWriter := io.Pipe()
+		catFileBatchReader, catFileBatchWriter := io.Pipe()
+		errChan := make(chan error, 1)
+		wg := sync.WaitGroup{}
+		wg.Add(6)
+		go func() {
+			defer wg.Done()
+			defer catFileBatchReader.Close()
+
+			bufferedReader := bufio.NewReader(catFileBatchReader)
+			buf := make([]byte, 1025)
+			for {
+				// File descriptor line
+				_, err := bufferedReader.ReadString(' ')
+				if err != nil {
+					catFileBatchReader.CloseWithError(err)
+					break
+				}
+				// Throw away the blob
+				if _, err := bufferedReader.ReadString(' '); err != nil {
+					catFileBatchReader.CloseWithError(err)
+					break
+				}
+				sizeStr, err := bufferedReader.ReadString('\n')
+				if err != nil {
+					catFileBatchReader.CloseWithError(err)
+					break
+				}
+				size, err := strconv.Atoi(sizeStr[:len(sizeStr)-1])
+				if err != nil {
+					catFileBatchReader.CloseWithError(err)
+					break
+				}
+				pointerBuf := buf[:size+1]
+				if _, err := io.ReadFull(bufferedReader, pointerBuf); err != nil {
+					catFileBatchReader.CloseWithError(err)
+					break
+				}
+				pointerBuf = pointerBuf[:size]
+				// now we need to check if the pointerBuf is an LFS pointer
+				pointer := lfs.IsPointerFile(&pointerBuf)
+				if pointer == nil {
+					continue
+				}
+				pointer.RepositoryID = pr.BaseRepoID
```

Comment:
One thing I'm not doing here is checking that the LFS object is present in the head repo. This only checks if the LFS object is present in the store.

I could add this check if wanted.

Reply:
I've added this check.